### DISCUSSION TITLE
CCACHE_DIRECT_ONLY option

### DISCUSSION THREAD

**Discussion by hkrpavan** [ccache/ccache#339] [General]
Do we have any option like not to do PREPROCESSOR mode?  I want to avoid "non direct" hits.

I am using distcc in pump mode. Usually full compilations happen on distcc server. I have patch working to use ccache in pump mode on distcc servers. So don't want to spend much time checking for preprocess hit locally. If there no direct cache hit just forward the compilation to distcc server.

Let me know if i can get for this functionality or else let me know which file in ccache sources to be looked in to? It saves lot of time for us.

**Comment by afbjorklund** [comment]
Currently we _have_ to run the preprocessor in order to get the original hash values, but there is some work done to use the depend files instead (#301). Then it would be interesting to also have an option to only use the manifests (direct mode), and not do any local preprocessing (except the distcc `include_server`).

You can use ccache/distcc already today, but it does require some documentation still (#320). As far as I know it shouldn't require patching - but please do make a Pull Request (to either project) if you found any issues. Also need better documentation on how to set up distributed cache, using moxi/memcached.

**Comment by jrosdahl** [comment]
See also #283 for a similar question.

@afbjorklund writes:
> (except the distcc include_server).

Could you expand on this part? I'm not very familiar with distcc unfortunately.

**Comment by hkrpavan** [comment]
Thanks! How to pull the patch for #301 I don't seem to have permissions. 

> You can use ccache/distcc already today, but it does require some documentation still (#320). As far as I know it shouldn't require patching

This requires patching on distccd (distcc server). In pump mode distcc server dynamically creates random numbered file names for objs and depend files under defined temp directory. This is causing cache misses on distcc servers. So had to create a patch for distcc server to have the full compilation happen in a base directory and a random number directory name under it. This will be compensated with CCACHE_BASEDIR. This fixed problems with ccaches misses on distcc server.

**Comment by jrosdahl** [comment]
> Thanks! How to pull the patch for #301

One way is to download and apply https://github.com/ccache/ccache/pull/301.diff.

> I don't seem to have permissions.

How did you try to pull the patch?

**Comment by afbjorklund** [comment]
@jrosdahl: the distcc "pump" mode comes with a preprocessor (of sorts), similar to the directives-only/rewrite-includes options, that will analyze the source file and output the list of dependencies...

This is then sorted into system files (which are supposed to be on the server already) and local files (which are sent to the server as part of the distribution). Unfortunately it is written in python, and not in C.

https://github.com/distcc/distcc/tree/master/include_server

https://github.com/distcc/distcc/blob/master/doc/protocol-3.txt

**Comment by afbjorklund** [comment]
@hkrpavan : ah, ok. Don't use ccache on the distcc servers. It used to be that it sent preprocessed files (which couldn't be hashed anyway), and now it is more efficient to send the hash instead of all the files.

But just like running with the `pump` wrapper in one-off mode (from ccache), it should still be possible to do - if wanted. Much better than the current _"distcc's pump mode is not compatible with ccache."_ anyway...

**Comment by hkrpavan** [comment]
Thanks guys! This patch helped me a lot. Its a saving of 30 mins on over 12K files compilation. Hope to see these changes in main branch soon.

**Comment by rmatev** [comment]
Sorry to post on this closed issue.

@hkrpavan would you mind sharing the distccd patch you mentioned? I tried to use ccache on the distcc server and ran into the same issue that you describe.

> This requires patching on distccd (distcc server). In pump mode distcc server dynamically creates random numbered file names for objs and depend files under defined temp directory. This is causing cache misses on distcc servers. So had to create a patch for distcc server to have the full compilation happen in a base directory and a random number directory name under it. This will be compensated with CCACHE_BASEDIR. This fixed problems with ccaches misses on distcc server.

**Reply by FlorianHeigl** [reply]
@rmatev as far as I understand it's https://patch-diff.githubusercontent.com/raw/ccache/ccache/pull/301.diff
I haven't checked if it got committed (yet)

regarding the initial bit of this issue, I think the flags on CCACHE side to have more determinism are something like `CCACHE_INODECACHE`

**Comment by FlorianHeigl** [comment]
@rmatev did you ever find a workaround of your own? Have been reading for a few hours but it's really unclear which issues still exist and how to get the combination working the best.
